### ISSUE TITLE
Using an env var other than NODE_ENV? (PR-welcome)

### ISSUE THREAD

**Issue by jameswomack** [lorenwest/node-config#336]
Hey there! 😸 
I've gotten a lot of use out of config. Thank you for all of your hard work and generosity. 

There's but two things that keep my teams from being able to use config on every single project. They are:
1. Inability to use something other than `${NODE_ENV}-${NODE_APP_INSTANCE}` for selecting config files
2. Inability to use config in the browser

A solution to no. 1 already has precedent in `NODE_CONFIG_DIR`. If we had something called `NODE_ENV_VAR_NAME` that effectively defaulted to `NODE_ENV`, that'd solve it.

The reason we need this functionality is that our organization has a concept of `STACK` that is different than `NODE_ENV`. We cannot use `NODE_ENV` because `NODE_ENV` is used by many other modules and we need to set it to production in prod regardless of the current `STACK`.

Would you be open to merging this functionality in if I created a PR for it?

**Comment by lorenwest** [comment]
Hi James, 

Thank you for your kind words. I dislike adding configs for config for getting too meta and this issue has a 2 line workaound, but since it keeps cropping up if you were to issue a pull request for `$NODE_ENV_VAR_NAME`, I'd be open to merging it in.

As for running config in the browser, were you thinking of maybe building a grunt task that bundled the proper config files into a .js file or module, and got that to the browser for node-config to find?

I'd be willing to start a conversation about the right architecture for this if we have enough interest. My projects all have server-side components that feed configurations into the index.html, but I can see the value for server-less applications.

**Comment by markstos** [comment]
@jameswomack, when you say "work in the browser" could describe what you mind? Is it a a serverless design,  or a design where node-config is used both on the frontend and backend separately? 

The pattern I use is to have node-config on the backend, but one of my top level configuration keys is named `frontend`.  This gets published at /config.json and the frontend apps pulls that in and uses as it as needed. 

Regarding:

> 1. Inability to use something other than ${NODE_ENV}-${NODE_APP_INSTANCE} for selecting config files

Coudl this be solved by using a different directory and NODE_CONFIG_DIR *or* by creating a symlink to the name format that node-config expects?

**Comment by jameswomack** [comment]
@markstos 
> when you say "work in the browser" could describe what you mind? 
Config being used both server and client side. I've used the technique of passing a JSON version of the config to the client via various means. It'd be great to be able to use the same way of accessing the keypaths (`foo.get('bar.baz.qux')`) and be able to `require` in config like others modules in the client too.

> Coudl this be solved by using a different directory and NODE_CONFIG_DIR or by creating a symlink to the name format that node-config expects?
No. The issue is that I need `NODE_ENV` to be `'production'` and yet have config select and merge config files based on `STACK`.

**Comment by jameswomack** [comment]
> Thank you for your kind words. I dislike adding configs for config for getting too meta and this issue has a 2 line workaound, but since it keeps cropping up if you were to issue a pull request for $NODE_ENV_VAR_NAME, I'd be open to merging it in.

You're welcome & thank you, that's great.

> As for running config in the browser, were you thinking of maybe building a grunt task that bundled the proper config files into a .js file or module, and got that to the browser for node-config to find?

My thinking is that it'd work in bundlers such as Browserify and Webpack is such a way that node-config would attempt to ```require(`./${NODE_CONFIG_DIR}/${NODE_ENV}`)``` and we'd rely on the bundler's module resolution strategy to have that work. Later on, when the browser module loader system is complete, it could work with that but I don't think we're there yet.

**Comment by kmudrick** [comment]
I would love this @jameswomack

**Comment by CarsonBanov** [comment]
Sorry to necro this a bit, but this discussion was useful in Sept 2017.  As I understand it, the goal is to segment the configs by a variable other than NODE_ENV (or maybe by NODE_END _and_ this 
other variable).  That's what I'm trying to do as well.

I think it is possible to achieve what @jameswomack wants by using the existing `NODE_CONFIG_DIR` setting.  It's a workaround for sure, but :man_shrugging: The basic idea is to use NODE_CONFIG_DIR to segment by that other variable $STACK, while leaving $NODE_ENV to segment by dev/prod.

Set up directory structure to segment by $STACK
```
/config
  /stack1
    default.js
    production.js
  /stack2
    default.js
    production.js
```
Set up build task to change $STACK and $NODE_ENV as appropriate
```
"build:stack1:dev": "NODE_ENV=development NODE_CONFIG_DIR=./config/stack1 your-build-task-here",
"build:stack1:prod": "NODE_ENV=production NODE_CONFIG_DIR=./config/stack1 your-build-task-here",
"build:stack2:prod": "NODE_ENV=production NODE_CONFIG_DIR=./config/stack2 your-build-task-here"
```
In my case, `your-build-task-here` is just `webpack --progress`, but it can be anything.

.cc @jameswomack @kmudrick

**Comment by iMoses** [comment]
@markstos isn't it solved by `NODE_CONFIG_ENV`?

**Comment by markstos** [comment]
@iMoses Yes, this was solved in November, 2017 by #428
